README.md
# CopyNESW

## About

CopyNESW is a Windows application designed for interfacing with Kevin Horton's CopyNES, a Nintendo Entertainment System modified to
support dumping cartridges, saving and restoring battery-backed RAM, uploading programs to RAM/Flash cartridges, and real-time debugging.
Nearly all functionality of the original MS-DOS QBASIC client has been implemented (in one form or another).

This client can be run on Windows 2000, Windows XP, Windows Vista, and Windows 7, 32-bit and 64-bit. Both the original parallel port 
CopyNES and USB CopyNES are supported by this client; parallel port mode should work on both 32-bit and 64-bit versions of Windows, though 
USB support has not yet been tested in 64-bit mode and may not be available.

## Downloads

- [Original CopyNES Client](http://kevtris.org/Projects/copynes/copynes.zip), written in QBASIC by Kevin Horton.
- [CopyNESW v3.0](http://www.qmtpro.com/~nes/copynes/copynesw_30.zip), the current build of the Win32 CopyNES client (last updated *Fri, 26 May 2017 21:36:52 +0000*).
- [CopyNESW Sources](https://github.com/quietust/copynesw), for anyone interested in adding their own functionality.
- [FTD2xx Driver](http://www.qmtpro.com/~nes/copynes/usb_driver.zip), must be installed if you are using a USB CopyNES

## Instructions

1. Download the original QBASIC client for CopyNES and extract it to a convenient directory, being sure to restore the directory structure.
2. Download the CopyNESW client and extract it into the COPYNES directory, restoring subdirectories. You should be prompted to overwrite CNRAM.BIN - do so.
3. If you are using a parallel port CopyNES, run install_parport_driver.exe - if you are running Windows Vista or Windows 7, you will be prompted to elevate. If you are using a USB CopyNES, download and install the above FTD2xx Driver if you have not already done so.
4. Run CopyNESW.exe. The menu should appear with only a small number of options enabled.
5. Click on the 'Options' button and select which port to use. If you are using a PCI parallel port card, you may need to specify a custom base address and ECP offset based on the I/O addresses it happens to use.
6. Upon pressing 'OK' (and each time you start the program), a version detection dialog should appear for about a second. If you are using an older CopyNES (with version 1 BIOS), the delay will be longer (3 seconds) and the 'Microbug' option will be disabled. If you are using parallel port mode and receive a warning about bidirectional communication, change your parallel port settings in your BIOS and enable ECP mode.
7. Read the CopyNES QBASIC documentation - a great majority of it is relevant to the Win32 version, and the rest should be fairly intuitive.

## Links

- Kevin Horton's [Official CopyNES page](http://kevtris.org/Projects/copynes/index.html)
- [USB CopyNES](http://www.retrousb.com/product_info.php?cPath=24&products_id=36), whose client is based on CopyNESW.
- [CopyNESW Homepage](http://www.qmtpro.com/~nes/copynes/)


NEWS.md
# News

## May 3, 2013

The source code for CopyNESW is now available on Github.

## November 10, 2012

Another minor update, this one actually loads plugin data from disk correctly and thus should actually work again.

## November 5, 2012

The download package for CopyNESW has been updated - the INI file now has correct initial paths, a parallel port I/O driver installer has been included (which will automatically prompt for elevation on Windows Vista and later), and it now includes the extra plugin binaries needed for additional RAM cart types.

Additionally, extra download links have been added for the original client (for convenience) and the USB driver (packaged separately because it's rather large and you only need to install it once), and the installation instructions have been updated.

## October 30, 2012

I've uploaded another new build, cleaning up the code a whole lot as well as fixing a bunch of bugs I introduced in the previous build.

## October 29, 2012

With timing totally concidental to my return to the NESdev community, I have released an updated version of CopyNESW which uses an updated inpout32.dll (to add support for Windows 7 64-bit), adds support for custom ECP port offsets (since my 64-bit driver uses base+0x8 instead of base+0x400), and incorporates all of the changes from bunnyboy's USB version to provide a unified client.

## February 4, 2007

As I have ceased involvement with the NESdev community, I have released the source code for the Win32 CopyNES client so that others can continue to update the client. A download link can be found below.

## July 12, 2006

The plugin select dialog has been updated to remember the last plugin selected.

## June 17, 2006

The CNRAM plugin has been updated to fix a bug in which it failed to unmap the CopyNES BIOS from memory.

## June 16, 2006

Added support for Memblers's UNROM Flash cartridge to the RAM Cart command.

Also made the RAM/Flash cartridge type selector use the global plugin select dialog.

## March 4, 2006

Support for custom parallel ports has been added - simply select the port name, then override the address in the configuration dialog. In addition, a program restart is no longer required when changing the port configuration.

I have implemented support for custom plugin categories - if you receive an error when attempting to use the "Run Plugin" or "Write WRAM" options, edit MAKEBASE.BAS (from the standard CopyNES distribution) and make the following changes:

Find the line DATA "*"," ",0,"Test & Dev Use".
Replace it with the following lines:
DATA "*"," ",2,"Test & Dev Use"
DATA "xxxx","xxxx.bin",998,"Custom Plugin"
Locate the "WRAM Readers" and "Nintendo: Famicom" groups.
Between them, insert the following category:
DATA "*"," ",1,"Save RAM Uploaders"
DATA "xxxx","xxxx.bin",998,"Custom writer"
DATA "NES-SxROM","sxram.bin",999,"SxROM (MMC1) SRAM writer"
DATA "NES-TxROM","txram.bin",999,"TxROM (MMC3) SRAM writer"
DATA "NES-ExROM","exram.bin",999,"ExROM (MMC5) SRAM writer"
DATA "NES-HxROM","hxram.bin",999,"HxROM (MMC6) SRAM writer"
Save the file and run it in QBASIC to recreate the file MAPPERS.DAT.
February 28, 2006

Support for the additional NetMos PCI parallel adapter I/O base addresses (E000 and E800) has been added (and the parallel port selection combobox has been resized to show all of them at once).

It is also now possible to unselect the parallel port and use the client offline (without having to wait for the version query to timeout).

## February 18, 2006

CopyNESW now features support for parallel ports operating in ECP mode.

In addition, it also now supports the NetMos High Speed PCI parallel adapter when using I/O base address D800 (so I can use my CopyNES with it).

## January 17, 2006

The VRC7 Tuner module has been implemented, and various bugs have been fixed in other modules (including filename selection in the 'Fix GAR' module).

## January 9, 2006

MicroBug is now mostly functional, though there are still some minor lockup bugs that need to be fixed.

## January 3, 2006

Fixed a bug where cancelling from a 'Select Plugin' dialog would crash the program.

I've also started work on MicroBug - currently, selecting the option will simply display the main dialog (without any functionality).

## January 1, 2006

My CopyNES arrived a few days ago, and I've had the opportunity to fix lots of bugs in the client. BankWatch now works properly, and 'RAM Cart' now supports two suboptions (NROM and CNROM, the latter being for a devcart I made).

'Dump Cart' still has a few issues to iron out (it can crash if you cancel at the wrong places), and 'Run Plugin' currently can only select WRAM writers (though fixing this requires a corresponding update to the QBASIC client).

My next priority will be implementing MicroBug, so keep an eye on this site for further updates.

## December 19, 2005

Added proper support for Custom plugin selection - when prompted to select a plugin which is not in the list, simply select "xxxx" from the first category.
You will then be prompted for a UNIF board name, an iNES mapper number, and the location of the plugin binary - for now, the binary MUST be located within the plugins directory (defaults to PLUGDONE).
This allows you to easily test new plugins without having to recreate MAPPERS.DAT.

Several bugs related to the plugin selection dialog have also been fixed.

## December 4, 2005

Created this page to hold the CopyNES Win32 client and relevant information.


src/config.cpp
#include "CopyNESW.h"

int	HWVer;
int	ParPort, ParAddr, ParECP;
BOOL	SaveCRC, SaveFiles, MakeUNIF;
char	Path_MAIN[MAX_PATH], Path_PRG[MAX_PATH], Path_CHR[MAX_PATH], Path_WRAM[MAX_PATH],
	Path_NES[MAX_PATH], Path_CRC[MAX_PATH], Path_NSF[MAX_PATH], Path_PLUG[MAX_PATH];

/* Note: base path must have a trailing backslash */
char *	_relpath (char *abspath, char *basepath)
{
	int i, j, k;
	char tmp[MAX_PATH];
	tmp[0] = 0;
	i = 0;
	
	while ((tolower(abspath[i]) == tolower(basepath[i])) && (basepath[i] != 0))
		i++;
	if (basepath[i])
	{
		k = 1;
		for (j = i; basepath[j] != 0; j++)
			if (basepath[j] == '\\')
				k++;
		for (j = 1; j < k; j++)
			strcat(tmp,"..\\");
	}
	strcat(tmp,&abspath[i]);
	if (strlen(tmp) == 0)
		strcpy(tmp,".");
	strcpy(abspath,tmp);
	return abspath;
}
char *	addSlash (char *path)
{
	int x = strlen(path);
	if (path[x-1] == '\\')
		return path;
	path[x] = '\\';
	path[x+1] = 0;
	return path;
}

void	GetProgPath (void)
{
	int i;
	GetModuleFileName(NULL,Path_MAIN,MAX_PATH);
	i = strlen(Path_MAIN);
	while (i > 0)
		if (Path_MAIN[--i] == '\\')
		{
			Path_MAIN[++i] = 0;
			return;
		}
}

void	GetConfig (void)
{
	char Config[MAX_PATH], tmpdir[MAX_PATH], tmpstr[16];
	strcpy(Config,Path_MAIN);
	strcat(Config,"CopyNESW.ini");
	SaveCRC = GetPrivateProfileInt("CopyNES","SaveCRC",0,Config);
	SaveFiles = GetPrivateProfileInt("CopyNES","SaveFiles",0,Config);
	MakeUNIF = GetPrivateProfileInt("CopyNES","MakeUNIF",0,Config);
	GetPrivateProfileString("CopyNES","PRGPath","Parts",Path_PRG,MAX_PATH,Config);
	GetPrivateProfileString("CopyNES","CHRPath","Parts",Path_CHR,MAX_PATH,Config);
	GetPrivateProfileString("CopyNES","WRAMPath","Finished",Path_WRAM,MAX_PATH,Config);
	GetPrivateProfileString("CopyNES","NESPath","Finished",Path_NES,MAX_PATH,Config);
	GetPrivateProfileString("CopyNES","CRCPath","CRC",Path_CRC,MAX_PATH,Config);
	GetPrivateProfileString("CopyNES","NSFPath","NSF",Path_NSF,MAX_PATH,Config);
	GetPrivateProfileString("CopyNES","PluginPath","Plugdone",Path_PLUG,MAX_PATH,Config);
	strcpy(Path_PRG,addSlash(_fullpath(strcpy(tmpdir,Path_MAIN),Path_PRG,MAX_PATH)));
	strcpy(Path_CHR,addSlash(_fullpath(strcpy(tmpdir,Path_MAIN),Path_CHR,MAX_PATH)));
	strcpy(Path_WRAM,addSlash(_fullpath(strcpy(tmpdir,Path_MAIN),Path_WRAM,MAX_PATH)));
	strcpy(Path_NES,addSlash(_fullpath(strcpy(tmpdir,Path_MAIN),Path_NES,MAX_PATH)));
	strcpy(Path_CRC,addSlash(_fullpath(strcpy(tmpdir,Path_MAIN),Path_CRC,MAX_PATH)));
	strcpy(Path_NSF,addSlash(_fullpath(strcpy(tmpdir,Path_MAIN),Path_NSF,MAX_PATH)));
	strcpy(Path_PLUG,addSlash(_fullpath(strcpy(tmpdir,Path_MAIN),Path_PLUG,MAX_PATH)));
	ParPort = GetPrivateProfileInt("CopyNES","ParPort",0,Config);
	GetPrivateProfileString("CopyNES","ParAddr","0",tmpstr,16,Config);
	sscanf(tmpstr,"%X",&ParAddr);
	GetPrivateProfileString("CopyNES","ParECP","0",tmpstr,16,Config);
	sscanf(tmpstr,"%X",&ParECP);
	if ((ParPort > 0) && (ParAddr == 0))
	{
		// update old config data
		if (ParPort == 1)
			ParAddr = 0x378;
		if (ParPort == 2)
			ParAddr = 0x278;
		if (ParPort == 3)
			ParAddr = 0x3BC;
		if ((ParPort == 4) || (ParPort == 9))
			ParAddr = 0xD800;
		if (ParPort == 5)
			ParAddr = 0xE000;
		if (ParPort == 6)
			ParAddr = 0xE800;
		if (ParPort > 4)
			ParPort = 4;
		if (ParECP == 0)
			ParECP = 0x400;
		WriteConfig();
	}
}
void	WriteConfig (void)
{
	char Config[MAX_PATH], tmpdir[MAX_PATH], tmpstr[16];
	strcpy(Config,Path_MAIN);
	strcat(Config,"CopyNESW.ini");
	sprintf(tmpstr,"%i",SaveCRC);
	WritePrivateProfileString("CopyNES","SaveCRC",tmpstr,Config);
	sprintf(tmpstr,"%i",SaveFiles);
	WritePrivateProfileString("CopyNES","SaveFiles",tmpstr,Config);
	sprintf(tmpstr,"%i",MakeUNIF);
	WritePrivateProfileString("CopyNES","MakeUNIF",tmpstr,Config);
	WritePrivateProfileString("CopyNES","PRGPath",_relpath(strcpy(tmpdir,Path_PRG),Path_MAIN),Config);
	WritePrivateProfileString("CopyNES","CHRPath",_relpath(strcpy(tmpdir,Path_CHR),Path_MAIN),Config);
	WritePrivateProfileString("CopyNES","WRAMPath",_relpath(strcpy(tmpdir,Path_WRAM),Path_MAIN),Config);
	WritePrivateProfileString("CopyNES","NESPath",_relpath(strcpy(tmpdir,Path_NES),Path_MAIN),Config);
	WritePrivateProfileString("CopyNES","CRCPath",_relpath(strcpy(tmpdir,Path_CRC),Path_MAIN),Config);
	WritePrivateProfileString("CopyNES","NSFPath",_relpath(strcpy(tmpdir,Path_NSF),Path_MAIN),Config);
	WritePrivateProfileString("CopyNES","PluginPath",_relpath(strcpy(tmpdir,Path_PLUG),Path_MAIN),Config);
	sprintf(tmpstr,"%i",ParPort);
	WritePrivateProfileString("CopyNES","ParPort",tmpstr,Config);
	sprintf(tmpstr,"%X",ParAddr);
	WritePrivateProfileString("CopyNES","ParAddr",tmpstr,Config);
	sprintf(tmpstr,"%X",ParECP);
	WritePrivateProfileString("CopyNES","ParECP",tmpstr,Config);
}

int	FindVersion (void)
{
	BYTE i;
	OpenStatus(topHWnd);
	StatusText("Querying CopyNES BIOS version...");
	if (!WriteByteEx(0xA2,3,FALSE))
	{
		StatusText("Failed to send version request!");
		StatusText("Make sure your CopyNES is connected and turned on!");
		StatusOK();
		return 0;	// write failed, device not present
	}
	StatusText("Waiting for reply...");
	if (!ReadByteEx(i,3,FALSE))
	{
		if (ParPort == -1)
		{
			StatusText("Version reply not received!");
			StatusText("Make sure your CopyNES is connected and turned on!");
			StatusOK();
			return 0;	// write failed, device not present
		}
		else
		{
			StatusText("Version reply not received! Assuming version 1 BIOS.");
			Sleep(SLEEP_LONG);
			CloseStatus();
			InitPort();
			ResetNES(RESET_COPYMODE);
			return 1;
		}
	}
	if ((i == 0xA2) && (ParPort != -1))
	{
		StatusText("Your parallel port does not support bidirectional communication!");
		StatusText("Please correct your BIOS settings and try again.");
		StatusOK();
		return 0;
	}
	StatusText("CopyNES identified as version %i.",i);
	Sleep(SLEEP_LONG);
	CloseStatus();
	// technically, these shouldn't be needed
	if (ParPort != -1)
		InitPort();
	ResetNES(RESET_COPYMODE);
	return i;
}

vector<Category *> Plugins;

static	void	trim (char *str)
{
	int i;
	for (i = strlen(str) - 1; i >= 0; i--)
	{
		if ((str[i] == ' ') || (str[i] == '\r') || (str[i] == '\n'))
			str[i] = 0;
		else	break;
	}
}

Plugin *makePlugin (const char *name, const char *filename, int num, const char *desc)
{
	// Special case for mapper 998 - return blank plugin
	if (num == 998)
	{
		Plugin *plugin = new Plugin;
		plugin->name = name;
		plugin->desc = desc;
		plugin->num = num;
		return plugin;
	}
	FILE *in = fopen(filename, "rb");
	if (!in)
	{
		in = fopen((string(Path_PLUG) + filename).c_str(), "rb");
		if (!in)
			return NULL;
	}
	fseek(in, 0, SEEK_END);
	int len = ftell(in);
	if (len < 1152)
	{
		fclose(in);
		return NULL;
	}
	fseek(in, 0, SEEK_SET);
	Plugin *plugin = new Plugin;
	plugin->name = name;
	plugin->desc = desc;
	plugin->num = num;
	fread(plugin->header, 1, 128, in);
	fread(plugin->data, 1, 1024, in);
	len -= 1152;
	// check for ROMstring
	if (len)
	{
		char *data = new char[len + 1];
		data[len] = 0;
		fread(data, 1, len, in);
		plugin->romstring = data;
		delete[] data;
	}
	fclose(in);
	return plugin;
}


BOOL	Startup	(void)
{
	char mapfile[MAX_PATH];
	FILE *PlugList;
	char *C1, *C2, *C3, *C4;

	InitCRC();
	GetProgPath();
	GetConfig();

	sprintf(mapfile,"%s%s",Path_MAIN, "mappers.dat");
	PlugList = fopen(mapfile, "rb");
	if (PlugList == NULL)
	{
		MessageBox(topHWnd,"Unable to open mappers.dat plugin list!", "CopyNES", MB_OK | MB_ICONERROR);
		return FALSE;
	}

	// step 1 - read header data

	// Header consists of length (in bytes) followed by column offsets (1-based) and category sizes
	char num[4];
	fread(num, 1, 4, PlugList);
	int header_len = atoi(num) - 4;

	int col0, col1, col2, col3, col4;

	char *header = new char[header_len + 1];
	fread(header, 1, header_len, PlugList);
	sscanf(header, "%i %i %i %i %i", &col0, &col1, &col2, &col3, &col4);
	free(header);

	C1 = new char[col1 - col0 + 1];	C1[col1 - col0] = 0;
	C2 = new char[col2 - col1 + 1];	C2[col2 - col1] = 0;
	C3 = new char[col3 - col2 + 1];	C3[col3 - col2] = 0;
	C4 = new char[col4 - col3 + 1];	C4[col4 - col3] = 0;
	
	// Next comes the actual plugin data

	Category *cat;
	Plugin *plug;
	while (!feof(PlugList))
	{
		fread(C1, 1, col1 - col0, PlugList);
		fread(C2, 1, col2 - col1, PlugList);
		fread(C3, 1, col3 - col2, PlugList);
		fread(C4, 1, col4 - col3, PlugList);
		trim(C1);
		trim(C2);
		trim(C3);
		trim(C4);
		if (!strcmp(C1,"*"))
		{
			if (!strcmp(C4,"end"))
				break;
			cat = new Category;
			cat->type = (plugin_type)atoi(C3);
			cat->desc = C4;
			Plugins.push_back(cat);
		}
		else
		{
			plug = makePlugin(C1, C2, atoi(C3), C4);
			if (plug)	cat->list.push_back(plug);
		}
	}

	fclose(PlugList);
	delete[] C1;
	delete[] C2;
	delete[] C3;
	delete[] C4;

	// RAMCART - create new category for upload plugins
	cat = new Category;
	cat->desc = "RAM/Flash cartridge programmer";
	cat->type = PLUG_UPLOAD;
	Plugins.push_back(cat);

	plug = makePlugin("NRAM", "ram.bin", 0, "NROM cart with 32K RAM for PRG and 8K RAM for CHR");
	if (plug)	cat->list.push_back(plug);

	plug = makePlugin("CNRAM", "cnram.bin", 1, "CNROM cart with 32K of PRG and CHR RAM");
	if (plug)	cat->list.push_back(plug);

	plug = makePlugin("UfROM", "uxram.bin", 2, "Memblers's flash cart for UNROM fun");
	if (plug)	cat->list.push_back(plug);

	plug = makePlugin("PowerPak Lite", "pplite.bin", 3, "PowerPak Lite RAM Cart loader");
	if (plug)	cat->list.push_back(plug);

	plug = makePlugin("PowerPak Boot", "pp.bin", 4, "PowerPak Boot Flasher");
	if (plug)	cat->list.push_back(plug);

	plug = makePlugin("Glider Flasher", "glider.bin", 5, "Glider House Flasher");
	if (plug)	cat->list.push_back(plug);
	// END RAMCART

	if (!OpenPort(ParPort, ParAddr, ParECP))
	{
		HWVer = 0;
		return TRUE;
	}
	if (ParPort != -1)
		InitPort();
	ResetNES(RESET_COPYMODE);
	Sleep(SLEEP_LONG);
	ResetNES(RESET_COPYMODE);
	HWVer = FindVersion();
	return TRUE;
}

BOOL	Shutdown (void)
{
	WriteConfig();
	while (Plugins.size())
	{
		Category *cat = Plugins.back();
		while (cat->list.size())
		{
			Plugin *plug = cat->list.back();
			delete plug;
			cat->list.pop_back();
		}
		delete cat;
		Plugins.pop_back();
	}
	ResetNES(RESET_PLAYMODE);
	ClosePort();
	return TRUE;
}
